### ISSUE TITLE
destination property setId() matches multiple source property hierarchies

### ISSUE THREAD

**Issue by jdhenckel** [modelmapper/modelmapper#239]
I don't know how to fix this.  I am using ModelMapper with all default settings. I tried STRICT and the problem goes away, but that causes other problems, so I cannot use STRICT.  Any advice?

The failure is on

    PostFocusAreaApproverDto dto = ....
    Test entity = modelMapper.map(dto, Test.class);

in which

    public class Test {
    
        private int id;
        private String createdByLanId;
        private Date createdDate;
        private Date endDate;
        private Date startDate;
        private String contactLanId;
    
        public int getId() {
            return id;
        }
        public void setId(int id) {
            this.id = id;
        }
        public String getCreatedByLanId() {
            return createdByLanId;
        }
        public void setCreatedByLanId(String createdByLanId) {
            this.createdByLanId = createdByLanId;
        }
        public Date getCreatedDate() {
            return createdDate;
        }
        public void setCreatedDate(Date createdDate) {
            this.createdDate = createdDate;
        }
        public Date getEndDate() {
            return endDate;
        }
        public void setEndDate(Date endDate) {
            this.endDate = endDate;
        }
        public Date getStartDate() {
            return startDate;
        }
        public void setStartDate(Date startDate) {
            this.startDate = startDate;
        }
        public String getContactLanId() {
            return contactLanId;
        }
        public void setContactLanId(String contactLanId) {
            this.contactLanId = contactLanId;
        }
    }

and also

    
    public class PostFocusAreaApproverDto {
        private String contactLanId;
        private Date endDate;
        private Date startDate;
        private int focusAreaId;
    
        public String getContactLanId() {
            return contactLanId;
        }
    
        public void setContactLanId(String contactLanId) {
            this.contactLanId = contactLanId;
        }
    
        public Date getEndDate() {
            return endDate;
        }
    
        public void setEndDate(Date endDate) {
            this.endDate = endDate;
        }
    
        public Date getStartDate() {
            return startDate;
        }
    
        public void setStartDate(Date startDate) {
            this.startDate = startDate;
        }
    
        public int getFocusAreaId() {
            return focusAreaId;
        }
    
        public void setFocusAreaId(int focusAreaId) {
            this.focusAreaId = focusAreaId;
        }
    }

THANKS!!

**Comment by jdhenckel** [comment]
actual error is

org.modelmapper.ConfigurationException: ModelMapper configuration errors:

The destination property spacetool.dto.Test.setId() matches multiple source property hierarchies:

	spacetool.dto.PostFocusAreaApproverDto.getFocusAreaId()
	spacetool.dto.PostFocusAreaApproverDto.getContactLanId()

**Comment by jdhenckel** [comment]
I found a workaround.  If I add a dummy method

    public int getId() { return 0; }

to the PostFocusAreaApproverDto then the error goes away.

**Comment by chhsiao90** [comment]
@jdhenckel - Thanks for the report and sorry for the lately reply, I had been busy recently...

Maybe you can try PropertyMap to skip the id field mapping, and please let me know that it works or not, thanks!
```java
    modelMapper.addMappings(new PropertyMap<PostFocusAreaApproverDto, Test>() {
      protected void configure() {
        skip().setId(0);
      }
    });
```

**Comment by jdhenckel** [comment]
This is a pretty good solution to my problem

  modelMapper.getConfiguration().setAmbiguityIgnored(true)

and the error goes away

**Comment by asinghi** [comment]
@jdhenckel This is working, but it's not a genuine solution. 
I have an entity which has Id, emailId, subscriptionId. While mapping Dto to domain, this error is occuring.

**Comment by jdhenckel** [comment]
@asinghi I made a[ post to SO]([url](https://stackoverflow.com/questions/44534172/how-to-customize-modelmapper)) about this

**Comment by johnluetke** [comment]
You should follow @chhsiao90's advice and create a PropertyMap for your source and destination DTOs. Using configuration options like STRICT and setAmbiguityIgnored will introduce other issues, as you have discovered.

**Comment by dharaoza** [comment]
I have the same problem can you please help to solve.
The destination property com.measureone.vo.OrgApplicationPerson.setPersonParty()/com.measureone.vo.Party.setModifiedByPersonParty() matches multiple source property hierarchies:

	com.measureone.dto.UserDTO.getModifiedByPersonPartyName()
	com.measureone.dto.UserDTO.getModifiedByPersonPartyId()

**Comment by chhsiao90** [comment]
I guessed that we need to improve our matching strategy that it looks like it didn't fulfill some situation, and make confused...

Currently, for similar situation maybe you can try
- Use explicit mapping and skip the destination properties
- Use setAmbiguityIgnored
- Use Strict Matching Strategy

**Comment by jdhenckel** [comment]
I would really help if you can provide a helper function such as this:  AddTwoWayMapping(Foo.class, Bar.class, Foo:getVehicleId, Bar:getVIN);   The one thing that automapper is missing is a simple "one liner" to add a two way mapping between certain fields in certain classes.  If this is possible, can you please post some sample code on this thread?

**Comment by HigorSenna** [comment]
Thank's for help @jdhenckel, i love you xD.
I used modelMapper.getConfiguration().setAmbiguityIgnored(true), and problem solved.

**Comment by JamieLui** [comment]
@jdhenckel  👍

**Comment by gotoark** [comment]
> modelMapper.getConfiguration().setAmbiguityIgnored(true)

This Resolves my Issues 💯

**Comment by TouchsopheaVireak-096** [comment]
I have the same problem can you please help to solve.
The destination property com.example.sb__jpa_reatjs_pgsql.ModelDTO.EmployeeDTO.setName() matches multiple source property hierarchies:
com.example.sb__jpa_reatjs_pgsql.model.Employee.getFirstName()
com.example.sb__jpa_reatjs_pgsql.model.Employee.getLastName()

**Comment by th1m0r** [comment]
Hi, about this issue. I have the same problem. And a solution proposed here, solve the problem, but i have a few considerations to avoid hidden bugs in the code.
I only to did was put, nothing more.

`modelMapper.getConfiguration().setAmbiguityIgnored(true);`

My entities model are"

`
class Section {
    private Long id;
}
class Group {
    private Long id
}
class Subgroup {
    private Long id
}
class Product {
    private Long id;
    private Section section;
    private Group group;
    private Subgroup subgroup;
}

class ProductInputDto {

    private Long idProduct;
    private Long idSection;
    private Long idGroup;
    private Long idSubgroup;

}
`

So, my concern is that configuration that solve de problem. but in future, change wrong values causing dificults bugs to trace.

Another problem, is that even creating a TypeMap, 
 the error occours. because before modelmapper uses the TypeMap, when do `modelMapper.map(product, ProductInputDto.class);` the exception occours not respecting a propertymap.

`modelMapper.createTypeMap(ProductInputDto.class,Product.class)
           .addMappings(mapper -> mapper.map(ProductInputDto::getProdutoId,Product::setId));`

So, the solution was setAmbiguityInored(true), but is the better solution?

**Comment by batu0b** [comment]
I can't say for the model mapper that this method works very well at the moment. The model mapping attempt for the Entity I created failed and gives the following error
**1) The destination property com.batu0b.blogApp.dto.commentDTO.GetCommentDTO.setId() matches multiple source property hierarchies:

	com.batu0b.blogApp.dto.commentDTO.CreateCommentDTO.getPostId()
	com.batu0b.blogApp.dto.commentDTO.CreateCommentDTO.getUserId()
**
_**My code is as follows**_ 
` @PostMapping
    public Comment create(@RequestBody @Valid CreateCommentDTO commentDTO) throws Exception {

        GetCommentDTO getCommentDTO = null;
        modelMapper.createTypeMap(CreateCommentDTO.class, GetCommentDTO.class).addMapping(CreateCommentDTO::getUserId, GetCommentDTO::setUserId).addMapping(CreateCommentDTO::getPostId, GetCommentDTO::setPostId).map(commentDTO, getCommentDTO);


        return null;
    }`
And my dto types are as follows

`@Data
@NoArgsConstructor
@AllArgsConstructor
public class CreateCommentDTO {
    @NotBlank
    private String text;
    @NotNull
    private Long postId;
    @NotNull
    private Long userId;
}
`
`@NoArgsConstructor
@AllArgsConstructor
@Data
public class GetCommentDTO {
    private Long id;
    private String text;
    private Long postId;
    private Long userId;
    private String userName;
    private Date createdAt;
    private Date updatedAt;
}
`
for some reason I can't make a map in this way, if you can help me I would be very grateful

**Comment by jo-ka** [comment]
I found another workaround, which I described in [this StackOverflow answer](https://stackoverflow.com/questions/49831753/modelmapper-matches-multiple-source-property-hierarchies/77732152#77732152).

Short summary: disable all the implicit mappings by setting up the `TypeMap` with `emptyTypeMap` and apply the required mappings explicitly. Maybe works in conjunction with `implicitMappings()` for the non-conflicting attributes, haven't tried that.
